### PR TITLE
net: lwm2m: call send timeout cb on client close

### PR DESCRIPTION
**Pull request by RomainPelletant**
Call timeout on registered send callback when client is closing
Issue #64289

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 1]
```diff
@@ -446,6 +446,10 @@ void lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 
 	for (i = 0, msg = messages; i < ARRAY_SIZE(messages); i++, msg++) {
 		if (msg->ctx == client_ctx) {
+			/* Timeout "send" method messages pending*/
+			if (msg->message_timeout_cb) {
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:450]: Wouldn't [send_status_cb()](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_object.h#L515C18-L515C32) with `LWM2M_SEND_STATUS_FAILURE` make more sense here? Timeout is not really what happens here.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 2]
```diff
@@ -1461,6 +1461,7 @@ int lwm2m_rd_client_stop(struct lwm2m_ctx *client_ctx,
 
 	client.ctx->event_cb = event_cb;
 	rd_client_message_free();
+	lwm2m_engine_context_close(client_ctx);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1464]: Context will be closed in `ENGINE_DEREGISTERED` state handler (which, depending on the `deregister` parameter is either entered immediately, or after the deregitration procedure), why would it be needed to do it here?

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 1 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:450]: You are right.
I suggest:
- Update static void do_send_timeout_cb(struct lwm2m_message *msg) to accept an enum as input argument

or 

- Update this function: static int do_send_reply_cb(const struct coap_packet *response, struct coap_reply *reply,
			    const struct sockaddr *from) to accept a specific code

Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:450]: But why not simply call `msg->send_status_cb(LWM2M_SEND_STATUS_FAILURE);` directly here?

Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c:450]: Sometimes I make life harder :)

**Comment by SeppoTakalo** [comment]
This is starting to be a bit blurry...
It is an user application that calls `lwm2m_rd_client_stop()`, and it will get `DISCONNECTED` event. So what do we achieve if we get yet another callback from send operation?

If the callback did not arrive, and application got DISCONNECTED event, then we know that ... actually, we don't know anything. Packet might, or might not have been received by the server.

If we need to be clear, that we have not sent the packet, we should check if the packet is still in `pending_sends` or `queued_messages` lists.
If it is not, it might have already been send, and might be received correctly by the server. So claiming that it is a FAILURE is not 100% correct.

**Comment by RomainPelletant** [comment]
@SeppoTakalo @rlubos thanks for your answers!
If _timeout_ or _failure_ are not appropriate, maybe an _unknown_ or _aborted_ could have sense in that case.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 2 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1464]: If device is more connected, deregistration will fail and block the closing process?
I can remove it from this PR: maybe out of scope. Please let me know.

**Comment by SeppoTakalo** [comment]
> If _timeout_ or _failure_ are not appropriate, maybe an _unknown_ or _aborted_ could have sense in that case.

I don't like to say that it does not make sense.. I'm just trying to say that we are not sure what the status of the packet is.

So maybe, could you refactor the PR so that if the packet is still on the `client_ctx->pending_sends` or
`client_ctx->queued_messages`, then we can say for sure that the send is failed.
If it is not on those queues, the packet has been send, and we have not yet received a confirmation for it.

I'm unsure of what would be the correct way. Claim to application that send has failed, when it might have passed, or to claim nothing if we are unsure. This feels like it should be handled in upper layer. For example  have timestampping on the data or something, to make sure that data that is send twice will not overlap.

**Comment by RomainPelletant** [comment]
@SeppoTakalo Yes, I will change to what you described.

- Check in pending or queued messages : if found in it => failed shall be called / If not found => I would suggest 5.03 Service Unavailable according to CoAP (maybe 5.04 Gateway Timeout?).

In my application, I use send method to upload data and the send response OK is flushing the data buffer. On error, the data is kept to be sent later.
In that case, unsure could be considered as error.

**Pull request closed without merging**